File: app/src/main/java/com/physphil/android/unitconverterultimate/models/Unit.java
/*
 * Copyright 2016 Phil Shadlyn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.physphil.android.unitconverterultimate.models;

import android.support.annotation.IntDef;

/**
 * A unit that can be converted to/from
 * Created by Phizz on 15-07-25.
 */
public class Unit {

    public static final int SQ_KILOMETRES = 0;
    public static final int SQ_METRES = 1;
    public static final int SQ_CENTIMETRES = 2;
    public static final int HECTARE = 3;
    public static final int SQ_MILE = 4;
    public static final int SQ_YARD = 5;
    public static final int SQ_FOOT = 6;
    public static final int SQ_INCH = 7;
    public static final int ACRE = 8;

    public static final int AUD = 1300;
    public static final int BGN = 1301;
    public static final int BRL = 1302;
    public static final int CDN = 1303;
    public static final int CHF = 1304;
    public static final int CNY = 1305;
    public static final int CZK = 1306;
    public static final int DKK = 1307;
    public static final int EUR = 1331;
    public static final int GBP = 1308;
    public static final int HKD = 1309;
    public static final int HRK = 1310;
    public static final int HUF = 1311;
    public static final int IDR = 1312;
    public static final int ILS = 1313;
    public static final int INR = 1314;
    public static final int JPY = 1315;
    public static final int KRW = 1316;
    public static final int MXN = 1317;
    public static final int MYR = 1318;
    public static final int NOK = 1319;
    public static final int NZD = 1320;
    public static final int PHP = 1321;
    public static final int PLN = 1322;
    public static final int RON = 1323;
    public static final int RUB = 1324;
    public static final int SEK = 1325;
    public static final int SGD = 1326;
    public static final int THB = 1327;
    public static final int LIRA = 1328;
    public static final int USD = 1329;
    public static final int ZAR = 1330;

    public static final int BIT = 100;
    public static final int BYTE = 101;
    public static final int KILOBIT = 102;
    public static final int KILOBYTE = 103;
    public static final int MEGABIT = 104;
    public static final int MEGABYTE = 105;
    public static final int GIGABIT = 106;
    public static final int GIGABYTE = 107;
    public static final int TERABIT = 108;
    public static final int TERABYTE = 109;

    public static final int JOULE = 200;
    public static final int KILOJOULE = 201;
    public static final int CALORIE = 208;
    public static final int KILOCALORIE = 202;
    public static final int BTU = 203;
    public static final int FT_LBF = 204;
    public static final int IN_LBF = 205;
    public static final int KILOWATT_HOUR = 206;
    public static final int ELECTRON_VOLT = 207;

    public static final int MPG_US = 300;
    public static final int MPG_UK = 301;
    public static final int L_100K = 302;
    public static final int KM_L = 303;
    public static final int MILES_L = 304;

    public static final int KILOMETRE = 400;
    public static final int MILE = 401;
    public static final int METRE = 402;
    public static final int CENTIMETRE = 403;
    public static final int MILLIMETRE = 404;
    public static final int MICROMETRE = 405;
    public static final int NANOMETRE = 406;
    public static final int YARD = 407;
    public static final int FEET = 408;
    public static final int INCH = 409;
    public static final int NAUTICAL_MILE = 410;
    public static final int FURLONG = 411;
    public static final int LIGHT_YEAR = 412;

    public static final int KILOGRAM = 500;
    public static final int POUND = 501;
    public static final int GRAM = 502;
    public static final int MILLIGRAM = 503;
    public static final int OUNCE = 504;
    public static final int GRAIN = 505;
    public static final int STONE = 506;
    public static final int METRIC_TON = 507;
    public static final int SHORT_TON = 508;
    public static final int LONG_TON = 509;

    public static final int WATT = 600;
    public static final int KILOWATT = 601;
    public static final int MEGAWATT = 602;
    public static final int HP = 603;
    public static final int HP_UK = 604;
    public static final int FT_LBF_S = 605;
    public static final int CALORIE_S = 606;
    public static final int BTU_S = 607;
    public static final int KVA = 608;

    public static final int MEGAPASCAL = 700;
    public static final int KILOPASCAL = 701;
    public static final int PASCAL = 702;
    public static final int BAR = 703;
    public static final int PSI = 704;
    public static final int PSF = 705;
    public static final int ATMOSPHERE = 706;
    public static final int TECHNICAL_ATMOSPHERE = 709;
    public static final int MMHG = 707;
    public static final int TORR = 708;

    public static final int KM_HR = 800;
    public static final int MPH = 801;
    public static final int M_S = 802;
    public static final int FPS = 803;
    public static final int KNOT = 804;

    public static final int CELSIUS = 900;
    public static final int FAHRENHEIT = 901;
    public static final int KELVIN = 902;
    public static final int RANKINE = 903;
    public static final int DELISLE = 904;
    public static final int NEWTON = 905;
    public static final int REAUMUR = 906;
    public static final int ROMER = 907;
    public static final int GAS_MARK = 908;

    public static final int YEAR = 1000;
    public static final int MONTH = 1001;
    public static final int WEEK = 1002;
    public static final int DAY = 1003;
    public static final int HOUR = 1004;
    public static final int MINUTE = 1005;
    public static final int SECOND = 1006;
    public static final int MILLISECOND = 1007;
    public static final int NANOSECOND = 1008;

    public static final int N_M = 1100;

    public static final int TEASPOON = 1200;
    public static final int TABLESPOON = 1201;
    public static final int CUP = 1202;
    public static final int FLUID_OUNCE = 1203;
    public static final int QUART = 1204;
    public static final int PINT = 1205;
    public static final int GALLON = 1206;
    public static final int BARREL = 1207;
    public static final int FLUID_OUNCE_UK = 1208;
    public static final int QUART_UK = 1209;
    public static final int PINT_UK = 1210;
    public static final int GALLON_UK = 1211;
    public static final int BARREL_UK = 1212;
    public static final int MILLILITRE = 1213;
    public static final int LITRE = 1214;
    public static final int CUBIC_CM = 1215;
    public static final int CUBIC_M = 1216;
    public static final int CUBIC_INCH = 1217;
    public static final int CUBIC_FOOT = 1218;
    public static final int CUBIC_YARD = 1219;

    @IntDef({SQ_KILOMETRES, SQ_METRES, SQ_CENTIMETRES, HECTARE, SQ_MILE, SQ_YARD, SQ_FOOT, SQ_INCH, ACRE,
            AUD, BGN, BRL, CDN, CHF, CNY, CZK, DKK, EUR, GBP, HKD, HRK, HUF, IDR, ILS, INR, JPY, KRW, MXN, MYR, NOK, NZD, PHP, PLN, RON, RUB, SEK, SGD, THB, LIRA, USD, ZAR,
            BIT, BYTE, KILOBIT, KILOBYTE, MEGABIT, MEGABYTE, GIGABIT, GIGABYTE, TERABIT, TERABYTE,
            JOULE, KILOJOULE, CALORIE, KILOCALORIE, BTU, FT_LBF, IN_LBF, KILOWATT_HOUR, ELECTRON_VOLT,
            MPG_US, MPG_UK, L_100K, KM_L, MILES_L,
            KILOMETRE, MILE, METRE, CENTIMETRE, MILLIMETRE, MICROMETRE, NANOMETRE, YARD, FEET, INCH, NAUTICAL_MILE, FURLONG, LIGHT_YEAR,
            KILOGRAM, POUND, GRAM, MILLIGRAM, OUNCE, GRAIN, STONE, METRIC_TON, SHORT_TON, LONG_TON,
            WATT, KILOWATT, MEGAWATT, HP, HP_UK, FT_LBF_S, CALORIE_S, BTU_S, KVA,
            MEGAPASCAL, KILOPASCAL, PASCAL, BAR, PSI, PSF, ATMOSPHERE, TECHNICAL_ATMOSPHERE, MMHG, TORR,
            KM_HR, MPH, M_S, FPS, KNOT,
            CELSIUS, FAHRENHEIT, KELVIN, RANKINE, DELISLE, NEWTON, REAUMUR, ROMER, GAS_MARK,
            YEAR, MONTH, WEEK, DAY, HOUR, MINUTE, SECOND, MILLISECOND, NANOSECOND,
            N_M,
            TEASPOON, TABLESPOON, CUP, FLUID_OUNCE, QUART, PINT, GALLON, BARREL, FLUID_OUNCE_UK, QUART_UK, PINT_UK,
            GALLON_UK, BARREL_UK, MILLILITRE, LITRE, CUBIC_CM, CUBIC_M, CUBIC_INCH, CUBIC_FOOT, CUBIC_YARD})
    public @interface id {}

    private int id;
    private int labelResource;
    private double conversionToBase;
    private double conversionFromBase;

    /**
     * Create a unit object
     *
     * @param id                 id of the unit
     * @param labelResource      string resource id of the label
     * @param conversionToBase   the value to convert to the base unit of the conversion
     * @param conversionFromBase the value to convert from the base unit of the conversion
     */
    public Unit(@id int id, int labelResource, double conversionToBase, double conversionFromBase) {
        this.id = id;
        this.labelResource = labelResource;
        this.conversionToBase = conversionToBase;
        this.conversionFromBase = conversionFromBase;
    }

    @id
    public int getId() {
        return id;
    }

    public double getConversionToBaseUnit() {
        return conversionToBase;
    }

    public double getConversionFromBaseUnit() {
        return conversionFromBase;
    }

    public int getLabelResource() {
        return labelResource;
    }
}

File: v6000/src/main/java/com/physphil/android/unitconverterultimate/models/Unit.kt
package com.physphil.android.unitconverterultimate.models

import com.physphil.android.unitconverterultimate.R

sealed class Unit {
    abstract val displayStringResId: Int
}

sealed class Area(override val displayStringResId: Int) : Unit() {
    object SqKilometre : Area(R.string.sq_kilometre)
    object SqMetre : Area(R.string.sq_metre)
    object SqCentimetre : Area(R.string.sq_centimetre)
    object Hectare : Area(R.string.hectare)
    object SqMile : Area(R.string.sq_mile)
    object SqYard : Area(R.string.sq_yard)
    object SqFoot : Area(R.string.sq_foot)
    object SqInch : Area(R.string.sq_inch)
    object Acre : Area(R.string.acre)

    companion object {
        val all: List<Area> = listOf(
            SqKilometre, SqMetre, SqCentimetre, Hectare, SqMile, SqYard, SqFoot, SqInch, Acre
        )
    }
}

sealed class Currency(
    val code: CurrencyCode,
    override val displayStringResId: Int
) : Unit() {
    object AustralianDollar : Currency(CurrencyCode.AUD, R.string.aud)
    object BulgarianLev : Currency(CurrencyCode.BGN, R.string.bgn)
    object BrazilianReal : Currency(CurrencyCode.BRL, R.string.brl)
    object CanadianDollar : Currency(CurrencyCode.CAD, R.string.cdn)
    object SwissFranc : Currency(CurrencyCode.CHF, R.string.chf)
    object ChineseYuan : Currency(CurrencyCode.CNY, R.string.cny)
    object CzechKoruna : Currency(CurrencyCode.CZK, R.string.czk)
    object DanishKrone : Currency(CurrencyCode.DKK, R.string.dkk)
    object Euro : Currency(CurrencyCode.EUR, R.string.eur)
    object BritishPound : Currency(CurrencyCode.GBP, R.string.gbp)
    object HongKongDollar : Currency(CurrencyCode.HKD, R.string.hkd)
    object CroatianKuna : Currency(CurrencyCode.HRK, R.string.hrk)
    object HungarianForint : Currency(CurrencyCode.HUF, R.string.huf)
    object IndonesianRupiah : Currency(CurrencyCode.IDR, R.string.idr)
    object IsraeliShekel : Currency(CurrencyCode.ILS, R.string.ils)
    object IndianRupee : Currency(CurrencyCode.INR, R.string.inr)
    object IcelandicKrona : Currency(CurrencyCode.ISK, R.string.isk)
    object JapaneseYen : Currency(CurrencyCode.JPY, R.string.jpy)
    object KoreanWon : Currency(CurrencyCode.KRW, R.string.krw)
    object MexicanPeso : Currency(CurrencyCode.MXN, R.string.mxn)
    object MalaysianRinggit : Currency(CurrencyCode.MYR, R.string.myr)
    object NorwegianKrone : Currency(CurrencyCode.NOK, R.string.nok)
    object NewZealandDollar : Currency(CurrencyCode.NZD, R.string.nzd)
    object PhilippinePeso : Currency(CurrencyCode.PHP, R.string.php)
    object PolishZloty : Currency(CurrencyCode.PLN, R.string.pln)
    object RomanianLeu : Currency(CurrencyCode.RON, R.string.ron)
    object RussianRouble : Currency(CurrencyCode.RUB, R.string.rub)
    object SwedishKrona : Currency(CurrencyCode.SEK, R.string.sek)
    object SingaporeDollar : Currency(CurrencyCode.SGD, R.string.sgd)
    object ThaiBaht : Currency(CurrencyCode.THB, R.string.thb)
    object TurkishLira : Currency(CurrencyCode.TRY, R.string.lira)
    object AmericanDollar : Currency(CurrencyCode.USD, R.string.usd)
    object SouthAfricanRand : Currency(CurrencyCode.ZAR, R.string.zar)

    companion object {
        val all: List<Currency> = listOf(
            AmericanDollar,
            AustralianDollar,
            BritishPound,
            BrazilianReal,
            BulgarianLev,
            CanadianDollar,
            ChineseYuan,
            CroatianKuna,
            CzechKoruna,
            DanishKrone,
            Euro,
            HongKongDollar,
            HungarianForint,
            IcelandicKrona,
            IndianRupee,
            IndonesianRupiah,
            IsraeliShekel,
            JapaneseYen,
            KoreanWon,
            MalaysianRinggit,
            MexicanPeso,
            NewZealandDollar,
            NorwegianKrone,
            PhilippinePeso,
            PolishZloty,
            RomanianLeu,
            RussianRouble,
            SingaporeDollar,
            SouthAfricanRand,
            SwedishKrona,
            SwissFranc,
            ThaiBaht,
            TurkishLira
        )
    }
}

sealed class DigitalStorage(override val displayStringResId: Int) : Unit() {
    object Bit : DigitalStorage(R.string.bit)
    object Byte : DigitalStorage(R.string.Byte)
    object Kilobit : DigitalStorage(R.string.kilobit)
    object Kilobyte : DigitalStorage(R.string.kilobyte)
    object Megabit : DigitalStorage(R.string.megabit)
    object Megabyte : DigitalStorage(R.string.megabyte)
    object Gigabit : DigitalStorage(R.string.gigabit)
    object Gigabyte : DigitalStorage(R.string.gigabyte)
    object Terabit : DigitalStorage(R.string.terabit)
    object Terabyte : DigitalStorage(R.string.terabyte)

    companion object {
        val all: List<DigitalStorage> = listOf(
            Bit, Byte, Kilobit, Kilobyte, Megabit, Megabyte, Gigabit, Gigabyte, Terabit, Terabyte
        )
    }
}

sealed class Energy(override val displayStringResId: Int) : Unit() {
    object Joule : Energy(R.string.joule)
    object Kilojoule : Energy(R.string.kilojoule)
    object Calorie : Energy(R.string.calorie)
    object Kilocalorie : Energy(R.string.kilocalorie)
    object Btu : Energy(R.string.btu)
    object FtLbF : Energy(R.string.ft_lbF)
    object InLbF : Energy(R.string.in_lbF)
    object KilowattHour : Energy(R.string.kilowatt_hour)

    companion object {
        val all: List<Energy> = listOf(
            Joule, Kilojoule, Calorie, Kilocalorie, Btu, FtLbF, InLbF, KilowattHour
        )
    }
}

sealed class Fuel(override val displayStringResId: Int) : Unit() {
    object MilesPerGallonUs : Fuel(R.string.mpg_us)
    object MilesPerGallonUk : Fuel(R.string.mpg_uk)
    object LitresPer100k : Fuel(R.string.l_100k)
    object KilometresPerLitre : Fuel(R.string.km_l)
    object MilesPerLitre : Fuel(R.string.miles_l)

    companion object {
        val all: List<Fuel> = listOf(
            MilesPerGallonUs, MilesPerGallonUk, LitresPer100k, KilometresPerLitre, MilesPerLitre
        )
    }
}

sealed class Length(override val displayStringResId: Int) : Unit() {
    object Kilometre : Length(R.string.kilometre)
    object Mile : Length(R.string.mile)
    object Metre : Length(R.string.metre)
    object Centimetre : Length(R.string.centimetre)
    object Millimetre : Length(R.string.millimetre)
    object Micrometre : Length(R.string.micrometre)
    object Nanometre : Length(R.string.nanometre)
    object Yard : Length(R.string.yard)
    object Feet : Length(R.string.feet)
    object Inch : Length(R.string.inch)
    object NauticalMile : Length(R.string.nautical_mile)
    object Furlong : Length(R.string.furlong)
    object LightYear : Length(R.string.light_year)

    companion object {
        val all: List<Length> = listOf(
            Kilometre, Mile, Metre, Centimetre, Millimetre, Micrometre, Nanometre, Yard, Feet, Inch, NauticalMile, Furlong, LightYear
        )
    }
}

sealed class Mass(override val displayStringResId: Int) : Unit() {
    object Kilogram : Mass(R.string.kilogram)
    object Pound : Mass(R.string.pound)
    object Gram : Mass(R.string.gram)
    object Milligram : Mass(R.string.milligram)
    object Ounce : Mass(R.string.ounce)
    object Grain : Mass(R.string.grain)
    object Stone : Mass(R.string.stone)
    object MetricTon : Mass(R.string.metric_ton)
    object ShortTon : Mass(R.string.short_ton)
    object LongTon : Mass(R.string.long_ton)

    companion object {
        val all: List<Mass> = listOf(
            Kilogram, Pound, Gram, Milligram, Ounce, Grain, Stone, MetricTon, ShortTon, LongTon
        )
    }
}

sealed class Power(override val displayStringResId: Int) : Unit() {
    object Watt : Power(R.string.watt)
    object Kilowatt : Power(R.string.kilowatt)
    object Megawatt : Power(R.string.megawatt)
    object Horsepower : Power(R.string.hp)
    object HorsepowerUk : Power(R.string.hp_uk)
    object FtLbFS : Power(R.string.ft_lbf_s)
    object CaloriePerSecond : Power(R.string.calorie_s)
    object BtuPerSecond : Power(R.string.btu_s)
    object Kva : Power(R.string.kva)

    companion object {
        val all: List<Power> = listOf(
            Watt, Kilowatt, Megawatt, Horsepower, HorsepowerUk, FtLbFS, CaloriePerSecond, BtuPerSecond, Kva
        )
    }
}

sealed class Pressure(override val displayStringResId: Int) : Unit() {
    object Megapascal : Pressure(R.string.megapascal)
    object Kilopascal : Pressure(R.string.kilopascal)
    object Pascal : Pressure(R.string.pascal)
    object Bar : Pressure(R.string.bar)
    object Psi : Pressure(R.string.psi)
    object Psf : Pressure(R.string.psf)
    object Atmosphere : Pressure(R.string.atmosphere)
    object TechnicalAtmosphere : Pressure(R.string.technical_atmosphere)
    object MmHg : Pressure(R.string.mmhg)
    object Torr : Pressure(R.string.torr)

    companion object {
        val all: List<Pressure> = listOf(
            Megapascal, Kilopascal, Pascal, Bar, Psi, Psf, Atmosphere, TechnicalAtmosphere, MmHg, Torr
        )
    }
}

sealed class Speed(override val displayStringResId: Int) : Unit() {
    object KilometresPerHour : Speed(R.string.km_h)
    object MilesPerHour : Speed(R.string.mph)
    object MetresPerSecond : Speed(R.string.m_s)
    object FeetPerSecond : Speed(R.string.fps)
    object Knot : Speed(R.string.knot)

    companion object {
        val all: List<Speed> = listOf(
            KilometresPerHour, MilesPerHour, MetresPerSecond, FeetPerSecond, Knot
        )
    }
}

sealed class Temperature(override val displayStringResId: Int) : Unit() {
    object Celsius : Temperature(R.string.celsius)
    object Fahrenheit : Temperature(R.string.fahrenheit)
    object Kelvin : Temperature(R.string.kelvin)
    object Rankine : Temperature(R.string.rankine)
    object Delisle : Temperature(R.string.delisle)
    object Newton : Temperature(R.string.newton)
    object Reaumur : Temperature(R.string.reaumur)
    object Romer : Temperature(R.string.romer)

    companion object {
        val all: List<Temperature> = listOf(
            Celsius, Fahrenheit, Kelvin, Rankine, Delisle, Newton, Reaumur, Romer
        )
    }
}

sealed class Time(override val displayStringResId: Int) : Unit() {
    object Year : Time(R.string.year)
    object Month : Time(R.string.month)
    object Week : Time(R.string.week)
    object Day : Time(R.string.day)
    object Hour : Time(R.string.hour)
    object Minute : Time(R.string.minute)
    object Second : Time(R.string.second)
    object Millisecond : Time(R.string.millisecond)
    object Nanosecond : Time(R.string.nanosecond)

    companion object {
        val all: List<Time> = listOf(
            Year, Month, Week, Day, Hour, Minute, Second, Millisecond, Nanosecond
        )
    }
}

sealed class Torque(override val displayStringResId: Int) : Unit() {
    object NewtonMetres : Torque(R.string.n_m)
    object FtLbF : Torque(R.string.ft_lbF)
    object InLbF : Torque(R.string.in_lbF)

    companion object {
        val all: List<Torque> = listOf(
            NewtonMetres, FtLbF, InLbF
        )
    }
}

sealed class Volume(override val displayStringResId: Int) : Unit() {
    object Teaspoon : Volume(R.string.teaspoon)
    object Tablespoon : Volume(R.string.tablespoon)
    object Cup : Volume(R.string.cup)
    object FluidOunce : Volume(R.string.fluid_ounce)
    object Quart : Volume(R.string.quart)
    object Pint : Volume(R.string.pint)
    object Gallon : Volume(R.string.gallon)
    object Barrel : Volume(R.string.barrel)
    object FluidOunceUk : Volume(R.string.fluid_ounce_uk)
    object QuartUk : Volume(R.string.quart_uk)
    object PintUk : Volume(R.string.pint_uk)
    object GallonUk : Volume(R.string.gallon_uk)
    object BarrelUk : Volume(R.string.barrel_uk)
    object Millilitre : Volume(R.string.millilitre)
    object Litre : Volume(R.string.litre)
    object CubicCentimetre : Volume(R.string.cubic_cm)
    object CubicMetre : Volume(R.string.cubic_m)
    object CubicInch : Volume(R.string.cubic_inch)
    object CubicFoot : Volume(R.string.cubic_foot)
    object CubicYard : Volume(R.string.cubic_yard)

    companion object {
        val all: List<Volume> = listOf(
            Teaspoon, Tablespoon, Cup, FluidOunce, Quart, Pint, Gallon, FluidOunceUk, QuartUk, PintUk, GallonUk, BarrelUk,
            Millilitre, Litre, CubicCentimetre, CubicMetre, CubicInch, CubicFoot, CubicYard
        )

        val cooking: List<Volume> = listOf(
            Teaspoon, Tablespoon, Cup, FluidOunce, FluidOunceUk, Pint, PintUk, Quart, QuartUk, Gallon, GallonUk, Millilitre, Litre
        )
    }
}